security/src/main/java/io/airlift/security/utils/KerberosUtil.java

```diff
@@ -0,0 +1,173 @@
+    {
+        KerberosConfiguration kerberosConfiguration = new KerberosConfiguration(subjectName, client);
+        String configName = client ? KerberosConfiguration.GSS_CLIENT : KerberosConfiguration.GSS_SERVER;
+        LoginContext context = new LoginContext(configName, null, null, kerberosConfiguration);
+        context.login();
+        return context.getSubject();
+    }
+
+    public static GSSContext getGssContext(Subject subject, String servicePrincipal, boolean initiator)
+    {
+        return Subject.doAs(subject, (PrivilegedAction<GSSContext>) () -> {
+            GSSContext context = null;
+            try {
+                Principal subjectPrincipal = subject.getPrincipals().iterator().next();
+                GSSManager gssManager = GSSManager.getInstance();
+
+                if (initiator) {
+                    GSSName gssUserName = gssManager.createName(subjectPrincipal.getName(), GSSName.NT_USER_NAME);
+                    GSSCredential clientCredential = gssManager.createCredential(gssUserName, GSSCredential.DEFAULT_LIFETIME,
+                            KERBEROS_OID, GSSCredential.INITIATE_ONLY);
+
+                    GSSName gssServiceName = gssManager.createName(servicePrincipal, null);
+                    context = gssManager.createContext(gssServiceName, SPNEGO_OID, clientCredential, GSSContext.INDEFINITE_LIFETIME);
+                }
+                else {
+                    GSSName gssServiceName = gssManager.createName(servicePrincipal, null);
+                    GSSCredential serverCredential = gssManager.createCredential(gssServiceName, GSSCredential.INDEFINITE_LIFETIME,
+                            SUPPORTED_OIDS, GSSCredential.ACCEPT_ONLY);
+                    context = gssManager.createContext(serverCredential);
+                }
+                context.requestMutualAuth(true);
+                context.requestConf(true);
+                context.requestInteg(true);
+                context.requestCredDeleg(false);
+                return context;
+            }
+            catch (GSSException e) {
+                throw Throwables.propagate(e);
+            }
+        });
+    }
+
+    public static class KerberosConfiguration
+            extends Configuration
+    {
+        public static final String GSS_SERVER = "gss-server";
+        public static final String GSS_CLIENT = "gss-client";
+        private static final Logger log = Logger.get(KerberosConfiguration.class);
+        private final String principal;
+
+        public KerberosConfiguration(String subjectName, boolean fromClient)
+                throws UnknownHostException
+        {
+            if (fromClient) {
+                this.principal = subjectName;
+            }
+            else {
+                this.principal = KerberosUtil.getServicePrincipal(subjectName, null);
+            }
+        }
+
+        @Override
+        public AppConfigurationEntry[] getAppConfigurationEntry(String name)
```

Comment:
this method will be called in the init() of LoginContext, which is for finding corresponding login module setting for specific application.